core/src/utils/prove.rs

```diff
@@ -164,15 +172,50 @@ where
             );
         }
     };
+    make_checkpoint_span.exit();
 
     // Commit to the shards.
     #[cfg(debug_assertions)]
     let mut debug_records: Vec<ExecutionRecord> = Vec::new();
+
+    let commit_span = tracing::debug_span!("Commit to shards").entered();
     let mut deferred = ExecutionRecord::new(program.clone().into());
     let mut state = public_values.reset();
     let nb_checkpoints = checkpoints.len();
     let mut challenger = prover.config().challenger();
     vk.observe_into(&mut challenger);
+
+    let (tx, rx) = sync_channel::<Vec<ExecutionRecord>>(
+        opts.shard_batch_size * opts.shard_chunking_multiplier,
+    );
+
+    let challenger_handle = {
+        let prover = prover.clone();
+        std::thread::spawn(move || {
+            let mut challenger = challenger;
+            for records in rx.iter().take(nb_checkpoints) {
+                // Save the public values for each shard.
+                let public_values = records
+                    .iter()
+                    .map(|record| {
+                        record.public_values::<SC::Val>()[0..prover.machine().num_pv_elts()]
+                            .to_vec()
+                    })
+                    .collect::<Vec<_>>();
```

Comment:
Not sure if this actually limits RAM usage, if prover is slow this could lead to a large record of records in a channel.

Reply:
Need to have some sort of rate limit